### ISSUE TITLE
Request: benchmarks on a compiled Linux kernel bzImage

### ISSUE THREAD

**Issue by joshtriplett** [divvun/bidiff#6]
Would you consider adding benchmarks on a compiled Linux kernel `bzImage` file? For instance, v5.3 "make defconfig" versus v5.4 "make defconfig"?

While I can test this locally easily enough, that wouldn't give results that can be compared with the other test cases in your benchmark table.

**Comment by fasterthanlime** [comment]
I'm not opposed to it, although those take a certain amount of time to set up properly and run..

Re comparing, what you're saying is that it would take me less time to run one more benchmark than it would take you to run the others on your machine?

I'm still kinda confused at to what kind of comparison you're looking for: I'd imagine the main concern would usually be the compression ratio, which would be identical on both machines - you could just run, say, zstd-ultra, brotli-q11, and a reasonable bsdiff preset and see if it's worth it?

(Again, I think it's interesting and I kinda want to to it, but I'd like to understand your potential use case better)

**Comment by joshtriplett** [comment]
@fasterthanlime 
> Re comparing, what you're saying is that it would take me less time to run one more benchmark than it would take you to run the others on your machine?

More to reproduce the test conditions for your benchmarks and feel confident that I hadn't missed something important. The benchmark description mentions the system you ran them on and the compression technique, but not (for instance) the specific command-line options for the `bic` tool to reproduce those results, or the benchmarking framework you used if you called the `bidiff` library directly.

> I'm still kinda confused at to what kind of comparison you're looking for:

Size and performance both. But the particular reason I asked after a Linux kernel `bzImage` is that it has a somewhat unusual format that doesn't normally diff well at all: it's a compiled ELF binary that embeds a compressed version of another compiled ELF binary. Most tools that operate on binaries, even those that try to understand executables and understand compressed files individually, don't tend to understand the combination of the two in the `bzImage` format well enough to do a good job unless they've specifically introduced support for that format.

Here's what I got with some local experimentation. I compared the v5.4 and v5.5 kernels, both built with `make -j$(nproc) defconfig` and then `make -j$(nproc)`; I saved both the top-level `vmlinux` (the uncompressed kernel) and `arch/x86/boot/bzImage` (the gzip-compressed kernel combined with the decompression stub). I used `bic diff --method zstd` on the old and new `vmlinux`, and on the old and new `bzImage`. I also used `zstd` (no arguments) to compress the new files.

Results:

```
~/src/linux$ du -b --apparent-size vmlinux.* bzImage.*
7511457	vmlinux.bic-patch
57665792	vmlinux.v5.4
58881248	vmlinux.v5.5
14204026	vmlinux.v5.5.zstd
8649250	bzImage.bic-patch
8868736	bzImage.v5.4
9016192	bzImage.v5.5
8762458	bzImage.v5.5.zstd
```

So, vmlinux diffs very well, while bzImage does not.

**Comment by fasterthanlime** [comment]
> So, vmlinux diffs very well, while bzImage does not.

I think your testing methodology is sound, and this is the expected result.

It's pretty much a given that bsdiff will perform poorly on high-entropy files, whether they're archives or something else.

I'm ready to bet that it finds almost no sequences in common, or that the sequences it finds have so many differences that it barely breaks even. It's probably very slow to diff, too!

If you were looking for a definitive "is bsdiff a safe bet for `bzImage` files", the answer is, no. And neither is any other compression format, I'm afraid!

**Comment by joshtriplett** [comment]
So, interestingly, I've found that `bidiff` works surprisingly well on other kinds of compressed files. Not quite as well as it would on the underlying uncompressed data, and there's still some data amplification, but still surprisingly well if a reasonably large amount of the uncompressed version is unchanged.

That's part of why I brought the issue up: I'm wondering if some enhancements to heuristics might potentially improve the results.

**Comment by bbqsrc** [comment]
I don't understand the premise. Binary diffing is, in essence, a form of compression. Applying it to already compressed data is a diminishing returns situation.

Use cases where it flourishes are large binaries that undergo small changes between versions, and where updating by redownloading it all would be a significant burden.

What is the use case you envisage for what you're suggesting?

**Comment by joshtriplett** [comment]
On Wed, Apr 08, 2020 at 02:45:20AM -0700, Brendan Molloy wrote:
> I don't understand the premise. Binary diffing is, in essence, a form of compression. Applying it to already compressed data is a diminishing returns situation.
> 
> Use cases where it flourishes are large binaries that undergo small changes between versions, and where updating by redownloading it all would be a significant burden.
> 
> What is the use case you envisage for what you're suggesting?

The uncompressed kernel code may have gone through a relatively small
change (e.g. the difference between x.y and x.y.1). A binary diff of the
uncompressed versions would be relatively small. But a binary diff of
the compressed versions doesn't manage to "see through" the compression,
even though it would be theoretically possible to do so.

The primary use case would be minimizing the size of updates.

**Comment by bbqsrc** [comment]
I don't mean to sound antagonistic, but isn't the solution here to just decompress the bzImage file first?

Introducing heuristics to essentially read through a number of compression formats seems like a lot of work for very little payoff, when the user knows in advance which compression is being used, and knows that a binary diff of the uncompressed format is going to pay much more significant dividends.

Given any updater requires knowing which file it's going to be using as the base of a patch in advance, any code that could be written to use bidiff (or any similar binary patcher) could take into account any necessary decompression and recompression steps that might be necessary.

**Comment by joshtriplett** [comment]
> isn't the solution here to just decompress the bzImage file first?

bzImage is a custom format (not something a normal compression tool extracts and can recreate), and also, doing that might not produce a binary-identical result.

In any case, the original purpose of this issue was to request the addition of a specific benchmark to the README. While I'd also love to see potential improvements for diffs on compressed data, that isn't the scope of this issue, and I wasn't intending for this issue to result in scope creep.

**Comment by bbqsrc** [comment]
Ah, thanks for the clarification, this was helpful. :smile:

**Comment by fasterthanlime** [comment]
Just to chime in again: I generally share @bbqsrc's assessment that diffing compressed data is a fool's errand and the only way to do it is to decompress the data first - and doing so as part of bidiff is way out of scope.

There's really nothing in bsdiff/bidiff heuristics that "knows about the contents of file". It performs well on uncompressed binaries because those tend to change by constant amounts in many places.

I understand that bzImage is a specific format, not just "vmlinux.bz2", I still think it can be done - it's just an ELF :) I'm sure the compressed payload is fairly easy to identify, but a naively "decompressed" version of it wouldn't be usable (it would contain all the other sections, including the decompressor, and the decompressed payload).

Again, that's a fascinating usecase, but way out of scope.

**Comment by joshtriplett** [comment]
@fasterthanlime Understood. And thank you.

I'd still be interested in seeing a benchmark included on a bzImage from (say) 5.6 to 5.6.3 stable, which will still show some improvement. (And people may end up doing that as part of a larger system image diff.)
